### PR TITLE
[70B-Part3] FSDP Training

### PR DESCRIPTION
**Pull request by farzadab**
This PR is mostly complete, but I do need to revisit some minor assumptions to make sure I didn't break the normal (non-fsdp) pipeline.

One of the odd changes here is the removal of `.to(device, dtype)`.
- The `dtype` part is moved into the model itself since it'll cost us too much to possibly be sloppy and load a full-precision 70B and then turn it into half precision.
- The `device` part is being handled by the `trainer` itself when doing `trainer.train`. This causes some minor issues (e.g. running `trainer.evaluate` before is not allowed), but they're not hugely important.
    - note that now `model.device` is `cpu` (or mps) before `trainer.train` and then `cuda:rank` after `trainer.train`. This might take some getting used to.

### CODE REVIEW

File: ultravox/training/train.py [hunk 1]
```diff
@@ -294,14 +290,19 @@ def train(args: config_base.TrainConfig):
             lr_scheduler_type=args.lr_scheduler,
             warmup_steps=args.lr_warmup_steps,
             weight_decay=args.weight_decay,
-            fp16=dtype == torch.float16,
-            bf16=dtype == torch.bfloat16,
+            # fp16=dtype == torch.float16,
+            # bf16=dtype == torch.bfloat16,
```
Comment (farzadab) [inline ultravox/training/train.py:294]: For some reason, when this is enabled, the `accelerate` library turns any non-FSDP weights to full precision which is not great for us. I could not find a way around this, but given that we could reproduce non-FSDP results exactly, I'm not worried about it for now.

I'll leave a TODO here to verify this on non-FSDP too.

Here's the piece of code that does this, including their explanation. If you can figure out why they're doing this, let me know 🙂 
image

Comment (farzadab) [reply ultravox/training/train.py:294]: Just verified that with all three PRs combined, the loss curve for non-FSDP workflow remains exactly the same. We The training speed also matches the previous version exactly.

File: ultravox/training/train.py [hunk 2]
```diff
@@ -310,17 +311,27 @@ def train(args: config_base.TrainConfig):
         logging.info("Starting training...")
         t_start = datetime.now()
         logging.info(f"train start time: {t_start}")
+
         if args.val_steps:
-            trainer.evaluate()
+            if args.use_fsdp:
```
Comment (juberti) [inline ultravox/training/train.py:316]: Can you explain what's happening here? Why isn't the model initialized with FSDP?

File: ultravox/model/ultravox_model.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ def __init__(self, config: UltravoxConfig):
 
         self.loss_config = LossConfig()
         self.post_init()
+        self.multi_modal_projector.to(dtype=config.torch_dtype)
```
[the remark this answers is not in the document]
Comment (farzadab) [reply ultravox/model/ultravox_model.py:59]: Right now, refactoring a function for just two lines of code seems a bit overkill to me. ZQ will likely create such a function with CFormer PR anyway, so maybe better to keep it out to avoid it getting removed in the merge?.

File: ultravox/training/config_base.py [hunk 4]
```diff
@@ -70,7 +73,7 @@ class TrainConfig:
     optimizer: str = "adamw_torch"
     num_epochs: int = 1
     max_steps: int = 0
-    val_steps: Optional[int] = None
+    val_steps: Optional[float] = None
```
[the remark this answers is not in the document]
Comment (farzadab) [reply ultravox/training/config_base.py:76]: `val_steps` (as well as `save_steps`) can be both int or float. I guess int is obvious. Float means percentage of training, so `val_steps = 0.25` means you'll have 4 checkpoints.

File: ultravox/training/train.py [hunk 5]
```diff
@@ -183,13 +185,7 @@ def train(args: config_base.TrainConfig):
...
     model.print_trainable_parameters()
 
-    # Move the model to GPU and enable bfloat16
-    dtype = getattr(torch, args.data_type)
-    device = torch.device(args.device, index=local_rank)
-    logging.info(
-        f"Using dtype and device (world_size): {dtype}, {device} ({world_size})"
-    )
-    model.to(device=device, dtype=dtype)
+    # TODO: move to device if not fsdp or will trainer do this?
```
[the remark this answers is not in the document]
Comment (farzadab) [reply ultravox/training/train.py:188]: Thanks for pointing it out. I kept the TODO to test something first. I did a test, but I just realized that I haven't verified that the initial validation works in the usual DDP mode.

File: ultravox/training/train.py (hunk 2 above)
Comment (farzadab) [reply ultravox/training/train.py:316]: In the above, we removed `.to(device)` in favour of transformers/accelerate doing that for us when wrapping the `model` object with `FSDP`.

That means the model is now on CPU which is not great if you want to run validation. It will be moved to GPU only after the `trainer.train` call.

We could wrap the model ourselves too if need be explicitly. Here's the code (we might be able to borrow `FullyShardedDataParallelPlugin` object to reuse the `set_auto_wrap_policy` function), but for now I had decided not to do that. Not a strong opinion either way.

image
image

File: ultravox/training/train.py [hunk 6]
```diff
@@ -310,17 +311,27 @@ def train(args: config_base.TrainConfig):
...
+                )
+            else:
+                trainer.evaluate()
+
         trainer.train()
         t_end = datetime.now()
         logging.info(f"train end time: {t_end}")
         logging.info(f"elapsed: {t_end - t_start}")
 
+    if args.use_fsdp:
```
[the remark this answers is not in the document]
Comment (farzadab) [reply ultravox/training/train.py:328]: From what I understand, we actually don't want `FULL_STATE_DICT` for model checkpoints that are created during training. That's why we set it to `SHARDED_STATE_DICT` above.
We do need this for doing the final save though.

I can't say this with 100% certainty though, since what people said would happen and what I saw happening in practice was different.

File: ultravox/training/train.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (farzadab) [reply ultravox/training/train.py]: Added back the `.to(device)` call when not using FSDP.

**Comment by farzadab** [comment]
@juberti are there any more comments?

File: ultravox/model/ultravox_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (juberti) [reply ultravox/model/ultravox_model.py:59]: maybe? Seems like it's also easy to merge a 2-line function, but I don't feel super strongly about this

File: ultravox/training/config_base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (juberti) [reply ultravox/training/config_base.py:78]: oh, huh, that's good to know. Suggest a comment to that effect for this variable.

File: ultravox/training/train.py (hunk 2 above)
Comment (juberti) [reply ultravox/training/train.py:314]: So... does this mean we can't run `evaluate` for fsdp models, unless we add manual wrapping? Were you planning to revisit this in a followup?

File: ultravox/training/train.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (juberti) [reply ultravox/training/train.py:326]: OK, this definitely seems like black magic stuff, please add a comment indicating this that we're changing the state dict back to non-sharded so it can be serialized properly.

**Review by juberti — APPROVED** [review]

File: ultravox/training/train.py (hunk 2 above)
Comment (farzadab) [reply ultravox/training/train.py:314]: `evaluate` (as in an evaluation that runs automatically after training) doesn't work either way now and I didn't want to spend time on it until I knew what ZQ wanted to do.

To clarify, the numbers I provided were using vllm as the inference engine.

File: ultravox/model/ultravox_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (farzadab) [reply ultravox/model/ultravox_model.py:59]: Done.
